File: problems/0003_longest-substring-without-repeating-characters/index-of/Solution.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        let chars = s.utf8CString
        var (m, start, end) = (0, 0, 0)
        while end < s.length {
            let idx = indexOf(chars, chars[end], start, end)
            end += 1
            if idx == nil {
                m = max(m, end - start)
            } else {
                start = idx! + 1
            }
        }
        return m
    }
    @inline(__always)
    func indexOf(_ chars: ContiguousArray<CChar>, _ c: CChar, _ start: Int, _ end: Int) -> Int? {
        for i in start..<end {
            if chars[i] == c {
                return i
            }
        }
        return nil
    }
}

File: problems/0003_longest-substring-without-repeating-characters/index-of/solution.c
inline int indexOf(const char* s, const char c, const int start, const int end) {
    for (int i = start; i < end; ++i)
        if (*(s + i) == c)
            return i;
    return -1;
}

int lengthOfLongestSubstring(char * s) {
    int max = 0, start = 0, end = 0;
    for (char c = *s; c != '\0'; c = *(s + end)) {
        const int idx = indexOf(s, c, start, end++);
        if (idx == -1) {
            const int l = end - start;
            if (l > max)
                max = l;
        } else
            start = idx + 1;
    }
    return max;
}
